pkg/controller/operators/catalog/subscription/reconciler.go

```diff
@@ -0,0 +1,163 @@
+			out = next
+		case CatalogHealthState:
+			// Gather catalog health and transition state
+			var catalogHealth []operators.SubscriptionCatalogHealth
+			catalogHealth, err = c.catalogHealth(s.Subscription().GetNamespace())
+			if err != nil {
+				break
+			}
+
+			out, err = s.UpdateHealth(c.now(), c.subscriptionClient, catalogHealth...)
+		case SubscriptionExistsState:
+			if s == nil {
+				err = errors.New("cannot transition nil state")
+				break
+			}
+			if s.Subscription() == nil {
+				err = errors.New("cannot transition with nil subscription")
+				break
+			}
+
+			// Set up fresh state
+			next = NewCatalogHealthState(s)
+		default:
+			// Ignore all other typestates
+			return
+		}
+	}
+
+	return
+}
+
+// catalogHealth gets the health of catalogs that can affect Susbcriptions in the given namespace.
+// This means all catalogs in the given namespace, as well as any catalogs in the operator's global catalog namespace.
+func (c *catalogHealthReconciler) catalogHealth(namespace string) ([]operators.SubscriptionCatalogHealth, error) {
+	catalogs, err := c.catalogLister.CatalogSources(namespace).List(labels.Everything())
+	if err != nil {
+		return nil, err
+	}
+
+	if namespace != c.globalCatalogNamespace {
+		globals, err := c.catalogLister.CatalogSources(c.globalCatalogNamespace).List(labels.Everything())
+		if err != nil {
+			return nil, err
+		}
+
+		catalogs = append(catalogs, globals...)
+	}
+
+	catalogHealth := make([]operators.SubscriptionCatalogHealth, len(catalogs))
+	now := c.now()
+	var errs []error
+	for i, catalog := range catalogs {
+		h, err := c.health(now, catalog)
+		if err != nil {
+			errs = append(errs, err)
+			continue
+		}
+
+		// Prevent assignment when any error has been encountered since the results will be discarded
+		if errs == nil {
+			catalogHealth[i] = *h
+		}
+	}
+
+	if errs != nil {
+		// Assign meaningful zero value
+		catalogHealth = nil
+	}
+
+	return catalogHealth, utilerrors.NewAggregate(errs)
+}
+
+// health returns a SusbcriptionCatalogHealth for the given catalog with the given now.
+func (c *catalogHealthReconciler) health(now *metav1.Time, catalog *operators.CatalogSource) (*operators.SubscriptionCatalogHealth, error) {
+	healthy, err := c.healthy(catalog)
+	if err != nil {
+		return nil, err
+	}
+
+	ref, err := reference.GetReference(catalog)
+	if err != nil {
+		return nil, err
+	}
+
+	h := &operators.SubscriptionCatalogHealth{
+		CatalogSourceRef: ref,
+		// TODO: Should LastUpdated be set here, or at time of subscription update?
+		LastUpdated: now,
+		Healthy:     healthy,
+	}
+
+	return h, nil
+}
+
+func (c *catalogHealthReconciler) healthy(catalog *operators.CatalogSource) (bool, error) {
+	// TODO: remove conversion when reconciler takes internal type
+	var ext *v1alpha1.CatalogSource
+	err := scheme.Convert(catalog, ext, 0)
+	if err != nil {
+		return false, err
+	}
+
+	return c.registryReconcilerFactory.ReconcilerForSource(ext).CheckRegistryServer(ext)
```

Comment:
non-blocking: we should probably memoize `ReconcilerForSource` so that we only have to construct each reconciler once